src/Workspaces/Core/Portable/Workspace/Workspace_EventsLegacy.cs

```diff
@@ -0,0 +1,140 @@
+    public event EventHandler<DocumentEventArgs> DocumentOpened
+    {
+        add => AddEventHandler(value, WorkspaceEventType.DocumentOpened);
+        remove => RemoveEventHandler(value, WorkspaceEventType.DocumentOpened);
+    }
+
+    /// <summary>
+    /// An event that is fired when any <see cref="TextDocument"/> is opened in the editor.
+    /// </summary>
+    public event EventHandler<TextDocumentEventArgs> TextDocumentOpened
+    {
+        add => AddEventHandler(value, WorkspaceEventType.TextDocumentOpened);
+        remove => RemoveEventHandler(value, WorkspaceEventType.TextDocumentOpened);
+    }
+
+    /// <summary>
+    /// An event that is fired when a <see cref="Document"/> is closed in the editor.
+    /// </summary>
+    public event EventHandler<DocumentEventArgs> DocumentClosed
+    {
+        add => AddEventHandler(value, WorkspaceEventType.DocumentClosed);
+        remove => RemoveEventHandler(value, WorkspaceEventType.DocumentClosed);
+    }
+
+    /// <summary>
+    /// An event that is fired when any <see cref="TextDocument"/> is closed in the editor.
+    /// </summary>
+    public event EventHandler<TextDocumentEventArgs> TextDocumentClosed
+    {
+        add => AddEventHandler(value, WorkspaceEventType.TextDocumentClosed);
+        remove => RemoveEventHandler(value, WorkspaceEventType.TextDocumentClosed);
+    }
+
+    /// <summary>
+    /// An event that is fired when the active context document associated with a buffer 
+    /// changes.
+    /// </summary>
+    public event EventHandler<DocumentActiveContextChangedEventArgs> DocumentActiveContextChanged
+    {
+        add => AddEventHandler(value, WorkspaceEventType.DocumentActiveContextChanged);
+        remove => RemoveEventHandler(value, WorkspaceEventType.DocumentActiveContextChanged);
+    }
+
+    private void AddEventHandler<TEventArgs>(EventHandler<TEventArgs> eventHandler, WorkspaceEventType eventType)
+        where TEventArgs : EventArgs
+    {
+        // Require main thread on the callback as this is used from publicly exposed events
+        // and those callbacks may have main thread dependencies.
+        IDisposable? disposer = null;
+
+        lock (_gate)
+        {
+            if (_disposableEventHandlers.TryGetValue((eventHandler, eventType), out var adviseCountAndDisposer))
+            {
+                (var adviseCount, disposer) = adviseCountAndDisposer;
+
+                // If we already have a handler for this event type, update the map with the new advise count
+                _disposableEventHandlers[(eventHandler, eventType)] = (adviseCount + 1, disposer);
+            }
+        }
+
+        if (disposer == null)
+        {
+            disposer = RegisterHandler(eventType, (Action<EventArgs>)Handler, WorkspaceEventOptions.RequiresMainThreadOptions);
+
+            lock (_gate)
+            {
+                _disposableEventHandlers[(eventHandler, eventType)] = (AdviseCount: 1, disposer);
+            }
+        }
```

Comment:
Why did we jump out of the lock here? Since there's a theoretical race if you had two parallel subscriptions where the TryGetValue misses both times, and then hits this path and the write happens twice. In that case AdviseCount should be 2, not 1.

(I can't imagine this ever happening in practice.)

Reply:
I had concerns about calling RegisterHandler inside the lock, but looking a bit closer, maybe that concern was unnecessary. I'll change this to not transition into the lock multiple times and we can chat more about this tomorrow.

Reply:
@jasonmalinowski -- Does this look good. Can we chat tomorrow on any remaining concerns you may have?